File: bezier/src/org/brogan/bezier/WeldRegistry.java
package org.brogan.bezier;
import java.util.*;

public class WeldRegistry {
    private final Map<CubicPoint, Set<CubicPoint>> links = new HashMap<>();

    public void registerWeld(CubicPoint a, CubicPoint b) {
        links.computeIfAbsent(a, k -> new HashSet<>()).add(b);
        links.computeIfAbsent(b, k -> new HashSet<>()).add(a);
    }

    public Set<CubicPoint> getLinked(CubicPoint p) {
        Set<CubicPoint> r = links.get(p);
        return r != null ? Collections.unmodifiableSet(r) : Collections.emptySet();
    }

    public void unregisterPoint(CubicPoint p) {
        Set<CubicPoint> linked = links.remove(p);
        if (linked != null) {
            for (CubicPoint o : linked) {
                Set<CubicPoint> s = links.get(o);
                if (s != null) {
                    s.remove(p);
                    if (s.isEmpty()) links.remove(o);
                }
            }
        }
    }

    /** Remove the link between a and b only, leaving each point's other links intact. */
    public void unregisterLink(CubicPoint a, CubicPoint b) {
        Set<CubicPoint> sa = links.get(a);
        if (sa != null) { sa.remove(b); if (sa.isEmpty()) links.remove(a); }
        Set<CubicPoint> sb = links.get(b);
        if (sb != null) { sb.remove(a); if (sb.isEmpty()) links.remove(b); }
    }

    public void clear() { links.clear(); }

    /**
     * Returns the full link map for snapshot capture.
     * Each entry maps a CubicPoint to the set of points it is welded to.
     */
    public java.util.Set<java.util.Map.Entry<CubicPoint, Set<CubicPoint>>> getEntries() {
        return Collections.unmodifiableSet(links.entrySet());
    }
}

File: bezier_py/model/weld_registry.py
"""
WeldRegistry — tracks cross-manager point identity for welded/coincident points.
Port of WeldRegistry.java.
"""
from __future__ import annotations


class WeldRegistry:
    """
    Bidirectional weld link registry.
    Maps CubicPoint → set[CubicPoint] for all welded partners.
    """

    def __init__(self) -> None:
        self._links: dict = {}  # dict[CubicPoint, set[CubicPoint]]

    def register_weld(self, a, b) -> None:
        """Mark a and b as welded to each other."""
        self._links.setdefault(a, set()).add(b)
        self._links.setdefault(b, set()).add(a)

    def get_linked(self, p) -> frozenset:
        """Return all points welded to p (empty frozenset if none)."""
        s = self._links.get(p)
        return frozenset(s) if s else frozenset()

    def unregister_point(self, p) -> None:
        """Remove p from the registry and clean up all reverse links."""
        linked = self._links.pop(p, None)
        if linked:
            for other in linked:
                s = self._links.get(other)
                if s:
                    s.discard(p)
                    if not s:
                        del self._links[other]

    def unregister_link(self, a, b) -> None:
        """Remove only the link between a and b, leaving other links intact."""
        sa = self._links.get(a)
        if sa:
            sa.discard(b)
            if not sa:
                del self._links[a]
        sb = self._links.get(b)
        if sb:
            sb.discard(a)
            if not sb:
                del self._links[b]

    def clear(self) -> None:
        self._links.clear()

    def entries(self):
        """Return items view of the link dict (for snapshot capture)."""
        return list(self._links.items())
